### DISCUSSION TITLE
Help with SQLAlchemy Connection Error in FastAPI

### DISCUSSION THREAD

**Discussion by Anderson-Andre-P** [fastapi/fastapi#12273] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from sqlalchemy.ext.declarative import declarative_base
from urllib.parse import quote

username = "anderson"
password = "bEA3@29&"
host = "localhost"
port = 3306
db_name = "ChallengeApplication"

escaped_password = quote(password)
URL_DATABASE = f'mysql+pymysql://{username}:{escaped_password}@{host}/{db_name}?port={port}?charset=utf8'
print("URL de Conexão:", URL_DATABASE)

engine = create_engine(URL_DATABASE)

try:
    connection = engine.connect()
    print("Conexão com SQLAlchemy bem-sucedida!")
except Exception as e:
    print("Erro ao conectar com SQLAlchemy:", e)
finally:
    if 'connection' in locals() and connection:
        connection.close()
```

### Description

Hello, I'm currently facing an issue when trying to connect my FastAPI application to a MySQL database using SQLAlchemy. I'm encountering the following error:

**Additional Information:**

I'm able to connect to the MySQL database using MySQL Workbench with the same credentials.
MySQL server is running, and I have verified the connection settings.
I have attempted to escape special characters in the password, but the error persists.

### Operating System

Linux

### Operating System Details

Ubuntu 22.04

### FastAPI Version

0.115.0

### Pydantic Version

2.9.2

### Python Version

 3.10.12

### Additional Context

I did it following a tutorial [TUTORIAL](https://www.youtube.com/watch?v=zzOwU41UjTM&t=1049s), but not even Chatgpt can solve it.

![image](https://github.com/user-attachments/assets/36edcceb-fea5-4986-8bd2-add6e6d4d88f)

**Comment by JavierSanchezCastro** [comment, marked as answer]
Not a FastAPI issue.

I think your URL is bad formatted, should be something like:

``mysql+pymysql://<user>:<password>@<host>:<port>/<db_name>``

You can use also MySQLDsn from pydantic to format your URL

**Reply by Anderson-Andre-P** [reply]
I made this correction, but the error persists.

`
username = "anderson"
password = "bEA3@29&"
host = "localhost"
port = 3306
db_name = "ChallengeApplication"

URL_DATABASE = f'mysql+pymysql://{username}:{password}@{host}:{port}/{db_name}'

`

![image](https://github.com/user-attachments/assets/6baabe32-36ba-4ae5-81be-ae9bb3f5403a)

**Comment by Anderson-Andre-P** [comment]
I found the error.

[Escaping Special Characters such as @ signs in Passwords](https://docs.sqlalchemy.org/en/20/core/engines.html#escaping-special-characters-such-as-signs-in-passwords)
